lib/services/provided/autocomplete.js

```diff
@@ -35,9 +35,8 @@ const regexes = {
 
 function parseCompletions(results: CompleteReply, prefix: string) {
   const { matches, metadata } = results;
-  let { cursor_start, cursor_end } = results;
-  cursor_start = char_idx_to_js_idx(cursor_start, prefix);
-  cursor_end = char_idx_to_js_idx(cursor_end, prefix);
+  const cursor_start = results.cursor_start;
+  const cursor_end = char_idx_to_js_idx(results.cursor_end, prefix);
```

Comment:
These changes would make `replacementPrefix` valid for completions within a line that includes unicode characters.
E.g.:
```
𝒓𝒂𝒏𝒅𝒏 = rand # `replacementPrefix` should be `rand`
```

Reply:
what do you mean replacement prefix should be rand? why is your cursor what do you have selected?

Reply:
The cursor is at the end of `rand` and try to complete to `rand` or `randn` or something.
On the `master`, the `replacementPrefix` would something like just `r`. Just give a try.

Reply:
what kernel? because i get no replacement

Reply:
i get no suggestions at all for rand

Reply:
i used `𝒓𝒂𝒏𝒅𝒏 = fil` and it correctly replaces for `file` and for `%%file`

Reply:
Ah, IJulia.
IPython would bring `match.start`  and `match.end`, thus it may not happen.

Reply:
Maybe the same issue would happen for any kernel other than IPython.

Reply:
Since I don't have another kernel I can't test this but can you do these test cases and verify.
The `|` denotes cursor, anything after this should not be included as a prefix. everything before until the space should.
I know there isn't a great way of testing, but do you believe these would pass.
`𝒓𝒂𝒏𝒅𝒏 = 𝒓𝒂nd|`
`𝒓𝒂𝒏𝒅𝒏 = 𝒓𝒂n|d`
`𝒓𝒂𝒏𝒅𝒏 = 𝒓𝒂|nd`
`𝒓𝒂𝒏𝒅𝒏 = 𝒓|𝒂nd`
Also are you sure that parsing cursor start is causing the problem. I mean is it literally return a different number like 1 instead of 0 using your test. Can you screenshot the scope variables in debug?

Reply:
I think no completion happens for `𝒓𝒂nd` by default in any language.
You mean bind `𝒓𝒂nd` to some function and try to complete ?

Reply:
https://jupyter-client.readthedocs.io/en/latest/messaging.html#completion
Read this carefully. Note the change in 5.2.

Reply:
We should remove the char to js parsing according to the docs. The must be a different logic error

Reply:
The problem may come from that no all the Jupyter kernels follow the 5.2 protocol.

Reply:
Than we either need a way to detect the protocol on kernel startup and save it in the kernel object. Or just support the latest protocol.

Reply:
Detecting protocol stuff doesn't sound good way, since it would make many branching and code will lead to be more complex.
To only support latest protocol isn't definitely sane approach. We lose support for many kernels, and it's destroying Hydrogen's beauty.

Rather I would like to take a regex-based approach, which is independent of `cursor_` feedback from kernels.

For now, I will restore `js_ind` stuff.

Reply:
I didn't think about regex. That could be interesting.

Reply:
Leaving `TODO` and revert back to the previous implementation for now.
The real fix will be by another PR.